# Support Vector Machines (SVMs)
In this notebook, we will learn a linear and kernalised method of SVMs, which can be used for both regression and classification. To start with, we will focus on binary classification. We will use stochastic gradient descent (SGD) for the optimisation of the hinge loss.

We will work with the [Breast Cancer Wisconsin (Diagnostic) Data Set](https://www.kaggle.com/uciml/breast-cancer-wisconsin-data), which you first need to download and then load in this notebook. If you faced difficulties downloading this data set from Kaggle, you should download the file directly from Blackboard. The data set contains various aspects of cell nuclei of breast screening images of patients with _(malignant)_ and without _(benign)_ breast cancer. Our goal is to build a classification model that can take these aspects of an unseen breast screening image, and classify it as either malignant or benign.

If you run this notebook locally on your machine, you will simply need to place the `csv` file in the same directory as this notebook.
If you run this notebook on Google Colab, you will need to use

  `from google.colab import files`

  `upload = files.upload()`

and then upload it from your local downloads directory.

In [60]:
# necessary imports
import numpy as np
import pandas as pd
import copy

In [2]:
 from google.colab import files

upload = files.upload()

Saving data.csv to data (1).csv


In [3]:
data = pd.read_csv('./data.csv')

# print shape and last 10 rows
print(data.shape)
data.tail(10)

(569, 33)


id diagnosis  ...  fractal_dimension_worst  Unnamed: 32
559  925291         B  ...                  0.08732          NaN
560  925292         B  ...                  0.08321          NaN
561  925311         B  ...                  0.05905          NaN
562  925622         M  ...                  0.14090          NaN
563  926125         M  ...                  0.09873          NaN
564  926424         M  ...                  0.07115          NaN
565  926682         M  ...                  0.06637          NaN
566  926954         M  ...                  0.07820          NaN
567  927241         M  ...                  0.12400          NaN
568   92751         B  ...                  0.07039          NaN

[10 rows x 33 columns]

We can see that our data set has 569 samples and 33 columns. The column `id` can be taken as an index for our pandas dataframe and `diagnosis` is the label (either **M: malignant** or **B: benign**).

Let's prepare the data set first of all by (i) cleaning it, (ii) separating label from features, and (iii) splitting it into train and test sets.

In [4]:
# drop last column (extra column added by pd)
data_1 = data.drop(data.columns[-1], axis=1)
# set column id as dataframe index
data_2 = data_1.set_index(data['id']).drop(data_1.columns[0], axis=1)

# check
data_2.tail()

diagnosis  radius_mean  ...  symmetry_worst  fractal_dimension_worst
id                             ...                                         
926424         M        21.56  ...          0.2060                  0.07115
926682         M        20.13  ...          0.2572                  0.06637
926954         M        16.60  ...          0.2218                  0.07820
927241         M        20.60  ...          0.4087                  0.12400
92751          B         7.76  ...          0.2871                  0.07039

[5 rows x 31 columns]

We do a bit more preparation by converting the categorical labels into 1 for **M** and -1 for **B**.

In [5]:
# convert categorical labels to numbers
diag_map = {'M': 1.0, 'B': -1.0}
data_2['diagnosis'] = data_2['diagnosis'].map(diag_map)

# put labels and features in different dataframes
y = data_2.loc[:, 'diagnosis']
X = data_2.iloc[:, 1:]

# check
print(y.tail())
X.tail()

id
926424    1.0
926682    1.0
926954    1.0
927241    1.0
92751    -1.0
Name: diagnosis, dtype: float64


radius_mean  texture_mean  ...  symmetry_worst  fractal_dimension_worst
id                                 ...                                         
926424        21.56         22.39  ...          0.2060                  0.07115
926682        20.13         28.25  ...          0.2572                  0.06637
926954        16.60         28.08  ...          0.2218                  0.07820
927241        20.60         29.33  ...          0.4087                  0.12400
92751          7.76         24.54  ...          0.2871                  0.07039

[5 rows x 30 columns]

As with any data set that has features over different ranges, it's required to standardise the data before.

In [6]:
## EDIT THIS FUNCTION
def standardise(X):
  mu = np.mean(X, 0)
  sigma = np.std(X, 0)
  X_std = (X - mu) / sigma ## <-- SOLUTION
  return X_std

In [7]:
X_std = standardise(X)

# check
X_std.tail()

radius_mean  texture_mean  ...  symmetry_worst  fractal_dimension_worst
id                                 ...                                         
926424     2.110995      0.721473  ...       -1.360158                -0.709091
926682     1.704854      2.085134  ...       -0.531855                -0.973978
926954     0.702284      2.045574  ...       -1.104549                -0.318409
927241     1.838341      2.336457  ...        1.919083                 2.219635
92751     -1.808401      1.221792  ...       -0.048138                -0.751207

[5 rows x 30 columns]

In [8]:
# insert 1 in every row for intercept b
X_std.insert(loc=len(X_std.columns), column='intercept', value=1)

# split into train and test set
# stacking data X and labels y into one matrix
data_split = np.hstack((X_std, y[:, np.newaxis]))

# shuffling the rows        
np.random.shuffle(data_split)

# we split train to test as 70:30
split_rate = 0.7
train, test = np.split(data_split, [int(split_rate*(data_split.shape[0]))])

X_train = train[:,:-1]
y_train = train[:, -1]

X_test = test[:,:-1]
y_test = test[:, -1]

y_train = y_train.astype(float)
y_test = y_test.astype(float)

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:6: FutureWarning: Support for multi-dimensional indexing (e.g. `obj[:, None]`) is deprecated and will be removed in a future version.  Convert to a numpy array before indexing instead.
  


## Linear SVM
We start with defining the hinge loss as
$$
\mathcal L (\boldsymbol w) = \frac{1}{2} \| \boldsymbol w \|^2 + \frac{\lambda}{n} \sum_{i=1}^n \max \bigg( 0, 1-y_i (\boldsymbol w \cdot x_i + b) \bigg) \, .
$$
where $\boldsymbol w$ is the vector of weights, $\lambda$ the regularisation parameter, and $b$ the intercept which is included in our `X` as an additional column of $1$'s.

In [9]:
# EDIT THIS FUNCTION
def compute_cost(W, X, y, regul_strength=1e5):
  n = X.shape[0]
  distances = 1 - y * (np.dot(X, W))  ## <-- SOLUTION
  distances[distances < 0] = 0  # equivalent to max(0, distance)
  hinge = regul_strength * (np.sum(distances) / n)  ## <-- SOLUTION

  # calculate cost
  cost = 1 / 2 * np.dot(W, W) + hinge
  return cost

Next, we need the gradients of this cost function.

In [10]:
# calculate gradient of cost
def calculate_cost_gradient(W, X_batch, y_batch, regul_strength=1e5):
  # if only one example is passed
  if type(y_batch) == np.float64:
      y_batch = np.asarray([y_batch])
      X_batch = np.asarray([X_batch])  # gives multidimensional array

  distance = 1 - (y_batch * np.dot(X_batch, W))
  dw = np.zeros(len(W))

  for ind, d in enumerate(distance):
      if max(0, d)==0:
          di = W
      else:
          di = W - (regul_strength * y_batch[ind] * X_batch[ind])
      dw += di

  dw = dw/len(y_batch)  # average
  return dw

Both of the two previous functions are then used in SGD to update the weights iteratively with a given learning rate $\alpha$. We also implement a stop criterion that ends the learning as soon as the cost function has not changed more than a manually determined percentage.

We know that the learning happens through updating the weights according to
$$
\boldsymbol w = \boldsymbol w - \alpha \frac{\partial \mathcal L}{\partial \boldsymbol w}
$$

where $\frac{\partial \mathcal L}{\partial \boldsymbol w}$ is the gradient of the hinge loss we have computed in the previous cell.

In [100]:
# EDIT THIS FUNCTION
def sgd(X, y, max_iterations=2000, stop_criterion=0.01, learning_rate=1e-5, regul_strength=1e5, print_outcome=False):
  # initialise zero weights
  weights = np.zeros(X.shape[1])
  nth = 0
  # initialise starting cost as infinity
  prev_cost = np.inf
  
  # stochastic gradient descent
  for iteration in range(1, max_iterations):
      # shuffle to prevent repeating update cycles
      np.random.shuffle([X, y])
      for ind, x in enumerate(X):
          ascent = calculate_cost_gradient(weights, x, y[ind], regul_strength) ## <-- SOLUTION
          weights = weights - (learning_rate * ascent)

      # convergence check on 2^n'th iteration
      if iteration==2**nth or iteration==max_iterations-1:
          # compute cost
          cost = compute_cost(weights, X, y, regul_strength)  ## <-- SOLUTION
          if print_outcome:
            print("Iteration is: {}, Cost is: {}".format(iteration, cost))
          # stop criterion
          if abs(prev_cost - cost) < stop_criterion * prev_cost:
              return weights
          
          prev_cost = cost
          nth += 1
  
  return weights

Now, we can take these functions and train a linear SVM with our training data.

In [101]:
# train the model
W = sgd(X_train, y_train, max_iterations=2000, stop_criterion=0.01, learning_rate=1e-3, regul_strength=1e3, print_outcome=True)
print("Training finished.")

Iteration is: 1, Cost is: 499.2965974291716
Iteration is: 2, Cost is: 638.5942711063158
Iteration is: 4, Cost is: 389.2082356070499
Iteration is: 8, Cost is: 611.3098658481113
Iteration is: 16, Cost is: 251.99622659053816
Iteration is: 32, Cost is: 301.7421196392772
Iteration is: 64, Cost is: 466.70496381255657
Iteration is: 128, Cost is: 562.309295997224
Iteration is: 256, Cost is: 296.892471277107
Iteration is: 512, Cost is: 241.00914001809392
Iteration is: 1024, Cost is: 533.393453002812
Iteration is: 1999, Cost is: 282.08448693574405
Training finished.


To evaluate the mean accuracy in both train and test set, we write a small function called `score`.

In [13]:
## EDIT THIS FUNCTION
def score(W, X, y):
  y_preds = np.array([])
  for i in range(X.shape[0]):
    y_pred = np.sign(np.dot(X[i], W))
    y_preds = np.append(y_preds, y_pred)
  
  return np.float(sum(y_preds==y)) / float(len(y)) ## <-- SOLUTION

In [14]:
print("Accuracy on train set: {}".format(score(W, X_train, y_train)))
print("Accuracy on test set: {}".format(score(W, X_test, y_test)))

Accuracy on train set: 0.9849246231155779
Accuracy on test set: 0.9415204678362573


#### Questions:
1. What are other evaluation metrices besides the accuracy? Implement them and assess the performance of our classification algorithm with them.
2. What makes other evaluation metrices more appropriate given our unbalanced data set _(we have more benign than malignant examples)_?
3. Try different learning rates, regularisation strengths and number of iterations independently. What can you observe? Can you achieve higher accuracies?
4. What is your understanding why have we used the hinge loss with this data set of 31 features? 
5. Can you think of other loss functions instead of the hinge loss? What is your intuition how they will perform compared to the hinge loss? You could try implementing one and compare the results. 

## *T*-fold cross validation

Now we repeat the same procedure as above but do not only have one train-test split, but multiple in a *T*-fold cross validation method.

In [102]:
def cross_val_split(data, num_folds):
  fold_size = int(len(data) / num_folds)
  data_perm = np.random.permutation(data)
  folds = []
  for k in range(num_folds):
    folds.append(data_perm[k*fold_size:(k+1)*fold_size, :])

  return folds

In [31]:
# evaluate
folds = cross_val_split(train, 5)

In [105]:
## EDIT THIS FUNCTION
def cross_val_evaluate(data, num_folds):
  
  folds = cross_val_split(data, num_folds)

  train_scores = []
  val_scores = []

  for i in range(len(folds)):
    print('Fold', i+1)
    # define the training set
    train_set = np.delete(np.asarray(folds).reshape(len(folds), folds[0].shape[0], folds[0].shape[1]), i, axis=0)
    train_folds = train_set.reshape(len(train_set)*train_set[0].shape[0], train_set[0].shape[1])
    X_train = train_folds[:,:-1]
    y_train = train_folds[:, -1]
    
    # define the validation set
    val_fold = folds[i]
    X_val = val_fold[:,:-1]
    y_val = val_fold[:, -1]

    # train the model
    W = sgd(X_train, y_train, max_iterations=1025, stop_criterion=0.01, learning_rate=1e-3, regul_strength=1e3)
    print("Training finished.")

    # evaluate
    train_score = score(W, X_train, y_train)
    val_score = score(W, X_val, y_val)
    print("Accuracy on train set #{}: {}".format(i+1, train_score))
    print("Accuracy on validation set #{}: {}".format(i+1, val_score))

    train_scores.append(train_score)
    val_scores.append(val_score)

  return train_scores, val_scores

In [106]:
train_scores, val_scores = cross_val_evaluate(train, 5)

Fold 1
Training finished.
Accuracy on train set #1: 0.990506329113924
Accuracy on validation set #1: 0.9620253164556962
Fold 2
Training finished.
Accuracy on train set #2: 0.990506329113924
Accuracy on validation set #2: 0.9746835443037974
Fold 3
Training finished.
Accuracy on train set #3: 0.9778481012658228
Accuracy on validation set #3: 1.0
Fold 4
Training finished.
Accuracy on train set #4: 0.9525316455696202
Accuracy on validation set #4: 0.9240506329113924
Fold 5
Training finished.
Accuracy on train set #5: 0.9841772151898734
Accuracy on validation set #5: 0.9873417721518988


Finally, let's compute the mean accuracy.

In [107]:
print(np.mean(val_scores))

0.969620253164557
